command/agent/exec/exec.go

```diff
@@ -0,0 +1,260 @@
+	processStateRunning
+	processStateRestarting
+	processStateStopped
+)
+
+type ServerConfig struct {
+	Logger      hclog.Logger
+	AgentConfig *config.Config
+
+	Namespace string
+
+	// LogLevel is needed to set the internal Consul Template Runner's log level
+	// to match the log level of Vault Agent. The internal Runner creates it's own
+	// logger and can't be set externally or copied from the Template Server.
+	//
+	// LogWriter is needed to initialize Consul Template's internal logger to use
+	// the same io.Writer that Vault Agent itself is using.
+	LogLevel  hclog.Level
+	LogWriter io.Writer
+}
+
+type Server struct {
+	// config holds the ServerConfig used to create it. It's passed along in other
+	// methods
+	config *ServerConfig
+
+	// runner is the consul-template runner
+	runner *manager.Runner
+
+	// numberOfTemplates is the count of templates determined by consul-template,
+	// we keep the value to ensure all templates have been rendered before
+	// starting the child process
+	// NOTE: each template may have more than one TemplateConfig, so the numbers may not match up
+	numberOfTemplates int
+
+	logger hclog.Logger
+
+	proc      *child.Child
+	procState processState
+
+	// exit channel of the child process
+	procExitCh <-chan int
+}
+
+type ProcessExitError struct {
+	ExitCode int
+}
+
+func (e *ProcessExitError) Error() string {
+	return fmt.Sprintf("process exited with %d", e.ExitCode)
+}
+
+func NewServer(cfg *ServerConfig) *Server {
+	server := Server{
+		logger:    cfg.Logger,
+		config:    cfg,
+		procState: processStateNotStarted,
+	}
+
+	return &server
+}
+
+func (s *Server) Run(ctx context.Context, incomingVaultToken chan string) error {
+	latestToken := new(string)
+	s.logger.Info("starting exec server")
+	defer func() {
+		s.logger.Info("exec server stopped")
+	}()
+
+	if len(s.config.AgentConfig.EnvTemplates) == 0 || s.config.AgentConfig.Exec == nil {
+		s.logger.Info("no env templates or exec config, exiting")
+		return nil
+	}
+
+	managerConfig := ctmanager.ManagerConfig{
+		AgentConfig: s.config.AgentConfig,
+		Namespace:   s.config.Namespace,
+		LogLevel:    s.config.LogLevel,
+		LogWriter:   s.config.LogWriter,
+	}
+
+	runnerConfig, err := ctmanager.NewConfig(managerConfig, s.config.AgentConfig.EnvTemplates)
+	if err != nil {
+		return fmt.Errorf("template server failed to generate runner config: %w", err)
+	}
+
+	// We leave this in "dry" mode, as there are no files to render;
+	// we will get the environment variables rendered contents from the incoming events
+	s.runner, err = manager.NewRunner(runnerConfig, true)
+	if err != nil {
+		return fmt.Errorf("template server failed to create: %w", err)
+	}
+
+	s.numberOfTemplates = len(s.runner.TemplateConfigMapping())
+
+	for {
+		select {
+		case <-ctx.Done():
+			s.runner.Stop()
+			if s.proc != nil {
+				s.proc.Stop()
+			}
+			s.procState = processStateStopped
+			return nil
+		case token := <-incomingVaultToken:
+			if token != *latestToken {
```

Comment:
Looking at this and thinking about it a bit, I'm wondering if this behaviour always makes sense. Consider the following application:

I want a short-lived auto-auth Vault token to grab 10 static secrets, then I want my app to run forever, I won't be able to support this workflow. I'd need to accept that my app will restart a bunch, or have an abnormally large token.

I'm wondering if we should have a similar config to our 'restart on new secret', which is something like 'restart on new token' (that we can default to 'always'). I'm interested to hear your thoughts, though this doesn't need to necessarily be changed as part of this PR, as it's a little bit of scope-creep, but I thought I'd bring it up as I was thinking about it.

Reply:
This is a very good point and could be a fairly common scenario.

If I understand correctly, the current code does not restart the child process on token changes (at least not directly) -- only the `runner` is restarted/reinitialized.

@VioletHynes do you know if we will get a new event on the `<-s.runner.TemplateRenderedCh():` when an expiring token triggers a dynamic secret refresh?

Reply:
>If I understand correctly, the current code does not restart the child process on token changes (at least not directly) -- only the runner is restarted/reinitialized.

Oh, you're totally right, I misread/misinterpreted this code! No need to worry about the use case I posted, I think.

>do you know if we will get a new event on the <-s.runner.TemplateRenderedCh(): when an expiring token triggers a dynamic secret refresh?

I don't know for sure, but I believe so. It's rendering a template, after all.